pkg/model/v1/decoder.go

```diff
@@ -0,0 +1,80 @@
+package v1
+
+import (
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/grafana/tempo/pkg/model/trace"
+	"github.com/grafana/tempo/pkg/tempopb"
+)
+
+const Encoding = "v1"
+
+type Encoder struct {
+}
+
+var staticEncoding = &Encoder{}
+
+func NewEncoding() *Encoder {
+	return staticEncoding
+}
+
+func (d *Encoder) Unmarshal(obj []byte) (*tempopb.Trace, error) {
+	trace := &tempopb.Trace{}
+	traceBytes := &tempopb.TraceBytes{}
+	err := proto.Unmarshal(obj, traceBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, bytes := range traceBytes.Traces {
+		innerTrace := &tempopb.Trace{}
+		err = proto.Unmarshal(bytes, innerTrace)
+		if err != nil {
+			return nil, err
+		}
+
+		trace.Batches = append(trace.Batches, innerTrace.Batches...)
+	}
+	return trace, err
+}
+
+func (d *Encoder) Matches(id []byte, obj []byte, req *tempopb.SearchRequest) (*tempopb.TraceSearchMetadata, error) {
+	t, err := d.Unmarshal(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	return trace.MatchesProto(id, t, req)
+}
+
+func (d *Encoder) Combine(objs ...[]byte) ([]byte, error) {
+	var combinedTrace *tempopb.Trace
+	for _, obj := range objs {
+		t, err := d.Unmarshal(obj)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshaling trace: %w", err)
+		}
+
+		combinedTrace, _ = trace.CombineTraceProtos(combinedTrace, t)
+	}
+
+	combinedBytes, err := d.Marshal(combinedTrace)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling combinedBytes: %w", err)
+	}
+
+	return combinedBytes, nil
+}
+
+func (d *Encoder) Marshal(t *tempopb.Trace) ([]byte, error) {
+	traceBytes := &tempopb.TraceBytes{}
+	bytes, err := proto.Marshal(t)
+	if err != nil {
+		return nil, err
+	}
+
+	traceBytes.Traces = append(traceBytes.Traces, bytes)
+
+	return proto.Marshal(traceBytes)
```

Comment:
Does this need to marshal to a buffer from `tempopb.SliceFromBytePool` ?  I don't think it does because the output buffer is not returned to the pool by the ingester, but double checking.

Reply:
tempopb.Trace doesn't use the `PreallocBytes` b/c it's just a thin wrapper around otel proto. `PushBytesRequest` is our only proto message tthat uses `PreallocBytes`.